grpclog: change logger to avoid Printf when the logger is io.Discard

rob05c opened the pull request:
Changes the logger to not call Printf when the logger is io.Discard.

Also changes the logger constructor to use `io.Discard` instead of `io.MultiWriter(io.Discard, io.Discard)`, which is necessary to detect all levels being discarded.

This is a significant performance improvement for disabled logs, see https://github.com/grpc/grpc-go/issues/7463 for details.

Notes:
- This is similar to how the std does it, see [log.go](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/log/log.go;l=203)
- I was hoping to use the std `log.Printf` itself, but I don't think that's possible, because the JSON option needs to Printf and then escape the result _after_ interpolating, which I don't think is possible with what's exposed by the std (possibly sans some overly complex io.Writer interceptors).
- It could use `log.Printf` for the non-JSON, but there's not much point if we can't use it for the JSON, and it would incur an unnecessary atomic call.
- The constructor `io.Discard` logic could be made faster (e.g. with more nested if's), and could do smarter things around the discard/multiwriter. I wrote it this way for readability, and because I don't think the performance is significant. But I'm happy to change it.

Fixes #7463

RELEASE NOTES:

- grpclog: Avoid expensive printf calls for io.Discard loggers.

purnesh42H commented on grpclog/loggerv2.go:
```diff
@@ -167,70 +190,89 @@ func newLoggerV2() LoggerV2 {
-	g.output(fatalLog, fmt.Sprintln(args...))
-	os.Exit(1)
+	g.println(fatalLog, args...)
+	os.Exit(FatalExitCode)
```
nit: I think 1 is a standard fatal exit code so we may not need a constant

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,220 @@ func checkLogForSeverity(s int, b []byte) error {
+						logger.Errorln("%v", msgErr)
+					}
+
+					checkWrite := func(buf *bytes.Buffer, discard bool, logType string, msg string, isJson bool) {
```
does checkWrite needs to be defined within the loop? Probably good to define all functions at the top

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,220 @@ func checkLogForSeverity(s int, b []byte) error {
+						bts, err := buf.ReadBytes('\n')
+						if discard {
+							if err == nil {
+								t.Fatalf("expected discard %v to not write, actual '%v'", logType, string(bts))
```
expected/actual can be changed to want/got. See here https://google.github.io/styleguide/go/decisions.html#got-before-want

purnesh42H commented on grpclog/loggerv2.go:
```diff
@@ -123,16 +123,37 @@ type loggerV2Config struct {
+
+	infoLogW := infoW
+
+	// use io.Discard instead of io.MultiWriter if all loggers at a given level are io.Discard.
```
nit: Please keep the comment lines upto 80 columns. Thanks

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2.go:
```diff
@@ -123,16 +123,37 @@ type loggerV2Config struct {
+
+	infoLogW := infoW
+
+	// use io.Discard instead of io.MultiWriter if all loggers at a given level are io.Discard.
```
Can we just check this condition of all loggers being io.Discard or not. If not, just do new assignments else don't. I think that will make the new code simpler

rob05c replied:
If we just check if all loggers are disabled, then if `Info` is disabled but `Warn` and `Err` aren't, then we won't discard `Info` logs. Or if `Info` and `Warn` are disabled but `Error` isn't, then we won't discard either `Info` or `Warn` logs.

So this would lose all the benefit unless someone has all logging entirely disabled.

Maybe I'm misunderstanding what you're asking?

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,231 @@ func checkLogForSeverity(s int, b []byte) error {
+		}
+	}
+
+	t.Run("discard only info", func(t *testing.T) {
```
why are these tests isolated outside of for loop. We can move them to new tests with explicit names specifying the test case

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,231 @@ func checkLogForSeverity(s int, b []byte) error {
 	return nil
 }
+
+func TestLoggerV2PrintFunc(t *testing.T) {
```
Can you convert the tests to table driven format like here [grpc-go/clientconn_test.go](https://github.com/grpc/grpc-go/blob/e6b6318ad90847fa8b5ed987b964f1494179c89f/clientconn_test.go#L811)

See https://pkg.go.dev/testing#hdr-Subtests_and_Sub_benchmarks for more details

rob05c replied:
I'm not entirely sure I understand what you're asking. I thought this was table-driven: https://github.com/grpc/grpc-go/pull/7471/files#diff-bac34975c1867d508c8aad68d2f029ec7935220afab764701a1c1f7d478be75eR190-R216

I changed it to call `t.Run` inside the loop, if that's what you mean?

purnesh42H replied:
Actually conventional way is to loop through flat list of test cases. Its okay if you have repetitions for some fields https://google.github.io/styleguide/go/decisions#table-driven-tests

rob05c replied:
Ah, ok. Changed.

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -42,7 +47,7 @@ func TestLoggerV2Severity(t *testing.T) {
 			b, err := buf.ReadBytes('\n')
 			if err != nil {
-				t.Fatal(err)
+				t.Fatalf("level %v: %v", j, err)
```
nit: since j is integer, you can just use `%d`

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+
+// makeLogFuncErrr returns a logger printFunc which generates a test error,
+// for testing that the log func is not called
+func makeLogFuncErr(t *testing.T) func() string {
```
can we move all the helper functions above the tests?

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+	checkLogContainsFuncStr(t, errB)
+}
+
+func TestLoggerV2PrintFuncAllDiscard(t *testing.T) {
```
what are we comparing in this test?

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+	bts, err := buf.ReadBytes('\n')
+	if discard {
+		if err == nil {
+			t.Fatalf("want discard %v to not write, got '%v'", logType, string(bts))
```
for comparison style guide recommends mentioning got before want https://google.github.io/styleguide/go/decisions#got-before-want

purnesh42H replied:
Please change to this pattern everywhere else as well

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+	}
+}
+
+// checkWrite checks that the log buffer buf was written as expected.
```
nit: the docstring is too verbose. My suggestion would be to be more explicit in naming the function and write inline comment where you are calling the function stating what you are trying to compare

rob05c replied:
Changed

purnesh42H commented on grpclog/loggerv2_test.go:
```diff
@@ -61,3 +66,264 @@ func checkLogForSeverity(s int, b []byte) error {
+	}
+}
+
+func TestLoggerV2Fatal(t *testing.T) {
```
this test looks different than others. Its hard to understand what are we verifying here

rob05c replied:
It's testing the Fatal functions. Because they call os.Exit, the only way to test them is to call them in a sub-process.
It's following https://go.dev/talks/2014/testing.slide#23

I added a comment explaining. 

I can remove them if you want. But then the Fatal functions won't be tested, and this won't have 100% test coverage.

easwars commented:
I'm sorry, but I merged a change recently that refactored the logging implementations. So, this PR might have significant merge conflicts now.

rob05c commented:
> I'm sorry, but I merged a change recently that refactored the logging implementations. So, this PR might have significant merge conflicts now.

I fixed the conflicts.

purnesh42H commented on grpclog/internal/loggerv2.go:
```diff
@@ -108,69 +108,88 @@ type loggerT struct {
 }
 
-func (g *loggerT) output(severity int, s string) {
+func (g *loggerT) printFunc(severity int, fn func() string) {
```
nit: probably need to be more explicit here with func name or docstring that it doesn't attempt to output the `fn` string in case of discard

rob05c replied:
I added a comment.

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -51,13 +126,243 @@ func TestLoggerV2Severity(t *testing.T) {
+	logger := NewLoggerV2(io.Discard, buffers[warningLog], buffers[errorLog], LoggerV2Config{})
+	loggerTp := logger.(*loggerT)
+
+	loggerTp.printFunc(infoLog, makeLogFuncErr(t))
```
probably a comment stating that `makeLogFuncErr` won't be called since infoLog is io.Discard?

purnesh42H replied:
Also, based on above NewLoggerV2(..) logic, since warningLog is not io.discard, infoLog shouldn't be either?

rob05c replied:
> probably a comment stating that makeLogFuncErr won't be called since infoLog is io.Discard?

Added

rob05c replied:
> Also, based on above NewLoggerV2(..) logic, since warningLog is not io.discard, infoLog shouldn't be either?

I'm not entirely sure what you mean?

If you mean the multi-writer logic:

```
	if warningW != io.Discard || infoW != io.Discard {
		warnLogW = io.MultiWriter(infoW, warningW)
	}
``` 

That's only setting `io.MultiWriter` on higher level logs. That is, `warn` is a MultiWriter for warn+info, but `info` is still a `io.Discard` (which is what this test is verifying).

If `logger.print(warnLog` is called, it will print to both `info` and `warn` (that behaviour predates this PR; this PR doesn't change any behaviour). So, when `info` is `io.Discard` and `warn` isn't, `fn` is called once, and the result written to the `io.MultiWriter` which will write to warn and info (which happens to be `io.Discard`). The `Printf` is only called once via `fn`, which is inevitable. (That behaviour is verified in `TestLoggerV2PrintFuncAllCombinations`.)

Does that answer your question?

purnesh42H replied:
Ah ok. In that case a docstring for test will be helpful. Something like this. 

```
TestLoggerV2PrintFuncDiscardOnlyInfo ensures that logs at the INFO level are discarded when set to io.Discard, while logs at other levels (WARN, ERROR) are still printed. It does this by using a custom error function that raises an error if the logger attempts to print at the INFO level, ensuring early return when io.Discard is used.
```

purnesh42H replied:
make sure to keep the comments cols upto 80 chars

rob05c replied:
Added

purnesh42H commented:
Overall looks good to me. Just one clarification needed about warnLog and infoLog which I asked in one of the above comment

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -51,13 +126,243 @@ func TestLoggerV2Severity(t *testing.T) {
+
+	// test that printFunc doesn't call the log func on discard loggers
+	// makeLogFuncErr will fail the test if it's called
+	loggerTp.printFunc(errorLog, makeLogFuncErr(t))
```
nit: we should check for other levels as well right to prevent future regressions?

rob05c replied:
Added

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -51,13 +126,243 @@ func TestLoggerV2Severity(t *testing.T) {
+	loggerTp.printFunc(errorLog, makeLogFuncErr(t))
+}
+
+func TestLoggerV2PrintFuncAllCombinations(t *testing.T) {
```
Do we really need to test all combinations? Based on the implementation we just want to verify the criticality based on verbosity i.e. if we will only allow io.Discard for current level if all the levels below it our io.Discard as well. For example, if warn is `io.Discard`, info needs to be io.Discard as well. I think the warn one is the only combination you are missing currently.

dfawley replied:
+1 for testing the combinations, but can we generate the table instead of manually enumerating all the cases? It should just be a few nested `for` loops?

rob05c replied:
Changed

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -20,11 +20,86 @@ package internal
 
+// check if b is in the format of:
+//
+//	2017/04/07 14:55:42 WARNING: WARNING
```
fix indenentation

rob05c replied:
I didn't add this, I just moved it to put helper functions at the top, like you requested previously.

The indentation here is being done by `gofmt`, if I remove it `gofmt` adds it back. Did you still want me to remove i?t

purnesh42H commented on grpclog/internal/loggerv2.go:
```diff
@@ -190,15 +211,36 @@ type LoggerV2Config struct {
+
+	infoLogW := infoW
+
+	// Use io.Discard instead of io.MultiWriter if all loggers at a given level
```
no need of new lines in between sentences. Suggestion;
```
// Use io.Discard instead of io.MultiWriter when all loggers at a given level
// are set to io.Discard. Both this package and the standard log package have
// significant optimizations for io.Discard, which io.MultiWriter lacks (as of this writing).
```

rob05c replied:
Changed

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -20,11 +20,86 @@ package internal
+	return nil
+}
+
+// makeLogFuncErrr returns a logger printFunc which generates a test error,
```
```
// makeLogFuncErr returns a logger printFunc that generates a test error.
// It is used to test that the log function is not called.
```

rob05c replied:
Changed

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -20,11 +20,86 @@ package internal
+
+// logFuncStr is a string used via testLogFuncOk and
+// testCheckLogContainsFuncStr to test the logger printFunc.
+const logFuncStr = "called-func"
```
const should be at the top

rob05c replied:
Moved

purnesh42H commented on grpclog/internal/loggerv2_test.go:
```diff
@@ -20,11 +20,86 @@ package internal
+// testCheckLogContainsFuncStr to test the logger printFunc.
+const logFuncStr = "called-func"
+
+// logFuncOk is a logger printFunc to write a constant string
```
```
// logFuncOk is a logger printFunc that returns a constant string to log.
```

rob05c replied:
Changed

purnesh42H commented:
Adding @dfawley as 2nd reviewer

rob05c commented:
Ok, I think I addressed all the comments again. Let me know if I missed anything.

Also, the PR was failing a new Github check for `RELEASE NOTES` in the description, so I attempted to add it like I see in other PRs. I'm happy to change it if it isn't right, just let me know

purnesh42H reviewed (approved):
lgtm

rob05c commented:
Is there anything else I need to do for this to be merged?

purnesh42H commented:
Doug is reviewing it. Nothing is needed from you right now.

dfawley commented on grpclog/internal/loggerv2.go:
```diff
@@ -190,15 +211,35 @@ type LoggerV2Config struct {
+	// significant optimizations for io.Discard, which io.MultiWriter lacks (as of
+	// this writing).
+
+	warnLogW := io.Discard
+	if warningW != io.Discard || infoW != io.Discard {
+		warnLogW = io.MultiWriter(infoW, warningW)
+	}
+
+	errLogW := io.Discard
+	if errorW != io.Discard || warningW != io.Discard || infoW != io.Discard {
+		errLogW = io.MultiWriter(infoW, warningW, errorW)
+	}
```
I think we could optimize this a bit, and make it easier to read as well?

```go
// Returns a combined logger for both higher & lower severity logs, or higher only if lower is io.Discard.
func combineLoggers(lower, higher io.Writer) io.Writer {
	if lower == io.Discard || higher == io.Discard {
		return higher
	}
	return io.MultiWriter(lower, higher)		
}

func NewLoggerV2(..) {
	// ...

	// Just overwrite instead of creating new locals with different names:
	warningW = combineLoggers(infoW, warningW)
	errorW = combineLoggers(warningW, errorW)
}
```

This ends up with 2 nested multiwriters if you have 3 different loggers, but that's probably fairly uncommon, and would be no worse than the behavior this PR adds of creating a multiwriter for `(warningW, io.Discard)` if `infoW` is `io.Discard`.

rob05c replied:
This would be a behavioural change. It's currently possible to set a lower level without a higher one, e.g. log info but not warning. With that change, any existing user doing that would find their info being discarded. Is that acceptable?

rob05c replied:
Ok, I chanaged it to:
```
if lower == io.Discard { 
	return higher 
} 
if lower == io.Discard { 
	return higher 
}
```

Which I believe preserves existing behaviour, if that still accomplishes what you want?

> // Just overwrite instead of creating new locals with different names:

I removed the locals, except `fatalW := errorW` to make the code self-documenting. But I can remove that too if you want.

> This ends up with 2 nested multiwriters if you have 3 different loggers

MultiWriter is smart enough to avoid this:
https://github.com/golang/go/blob/946b4baa/src/io/multi.go#L127-L131

dfawley reviewed (commented):
Thanks for the PR and sorry again for the delay.

rob05c commented:
> This PR is labeled as requiring an update from the reporter, and no update has been received after 6 days. If no update is provided in the next 7 days, this issue will be automatically closed.

Sorry, I was on vacation last week. I'll try to address the comments sometime this week

rob05c commented:
N.B. I made a few changes since the last review, adding

```
var sprintf = fmt.Sprintf
var exit = os.Exit
```

in loggerv2.go, to make it possible to test that `os.Exit` is called and `fmt.Sprintf` isn't. 

Testing Sprintf not being called on io.Discard was previously done via the closure. After removing the closure, this allows the test to change the var to test.

For the exit, the Github ARM runner started breaking with the external call this PR was previously doing to test log.Fatal/os.Exit (a la https://go.dev/talks/2014/testing.slide#23), `fork/exec /tmp/go-build692284400/b709/internal.test: invalid argument`. It seems that since a recent change the arm test is doing something strange with processes. I couldn't get it to work, so I changed to the var instead, which the test can change, so it can be tested in the same process without actually exiting.

IMO the tests are valuable, but I realise it's a bit unusual, I can remove the vars and corresponding tests if you don't agree.

dfawley reviewed (approved):
LGTM, thank you!  PTAL @purnesh42H
